New ai quickstarts (#39679)
dotnet/docs: docs/azure/ai/get-started-azure-openai.md

---

> title: Get started with the .NET and Azure Open AI

Reviewer: ```suggestion
title: Quickstart - Build an Azure AI chat app with .NET
```

description: Get started on how to use the Azure OpenAI with a gpt-35-turbo model, from a simple .NET console application. Get a hiking recommendation from the AI model. It consists of a simple console application, running locally, that will send request to an Azure OpenAI service deployed in your Azure subscription.

> ms.date: 02/20/2024

Reviewer: ```suggestion
ms.date: 02/28/2024
```

> ms.date: 02/20/2024

Reviewer: ```suggestion
description: Use the .NET Azure OpenAI SDK to build a simple .NET console chat application.
```

> ms.topic: get-started

Reviewer: check to make sure this is a valid ms.topic

Reviewer 2: get-started is present in the list https://review.learn.microsoft.com/en-us/help/platform/metadata-taxonomies?branch=main#mstopic

ms.custom: devx-track-dotnet, devx-track-dotnet-ai
author: fboucher
ms.author: frbouche
# CustomerIntent: As a .NET developer new to Azure OpenAI, I want deploy and use sample code to interact to learn from the sample code.
---

> This sample demonstrates how to use the Azure OpenAI with a `gpt-35-turbo` model, from a simple .NET 8.0 console application. It consists of a simple console application, running locally, that will send request to an Azure OpenAI service deployed in your Azure subscription. Everything will be deployed automatically using the Azure Developer CLI.

Reviewer: ```suggestion
Get started with the .NET Azure OpenAI SDK by creating a simple .NET 8 console chat application. The application will run locally and use the OpenAI `gpt-35-turbo` model deployed into an Azure OpenAI account. Follow these steps to provision Azure OpenAI and learn how to use the .NET Azure OpenAI SDK.
```

> ## Requirements

Reviewer: ```suggestion
## Prerequisites
```

> - .NET 8.0 SDK - [Install the .NET 8.0 SDK](https://dotnet.microsoft.com/download/dotnet/8.0)
> - An Azure subscription - [Create one for free](https://azure.microsoft.com/free/)
> - Azure Developer CLI - [Install or update the Azure Developer CLI](https://learn.microsoft.com/azure/developer/azure-developer-cli/install-azd)

Reviewer: ```suggestion
- .NET 8.0 SDK - [Install the .NET 8.0 SDK](https://dotnet.microsoft.com/download/dotnet/8.0)
- An Azure subscription - [Create one for free](https://azure.microsoft.com/free/)
- Azure Developer CLI - [Install or update the Azure Developer CLI](/azure/developer/azure-developer-cli/install-azd)
- Access to [Azure OpenAI service](/azure/ai-services/openai/overview#how-do-i-get-access-to-azure-openai) or a [OpenAI API access](https://openai.com/product).
```

> 

Reviewer: Let's make the pre-reqs an include file. Add an `include` folder under this directory and add everything under the prerequisites header to it. Then reference that.

Reviewer 2: Yes, good idea.
I started doing this also on the code side.

## Getting Started

Ensure that you follow the prerequisites to have access to Azure OpenAI Service as well as the Azure Developer CLI, and then follow the following guide to set started with the sample application.

1. Clone/ Download the repository: [dotnet/ai-samples](https://github.com/dotnet/ai-samples)
1. From a terminal or command prompt, navigate to the `01-HikerAI` directory.
1. Create the Azure resources using the Azure Developer CLI:
	```bash
	azd up
	```

> 1. It's now time to try the console application. Depending on your Azure subscription it's possible that a few (~5) minutes more minute are required before the model deployed in Azure OpenAI get available. If you get an error message about this, wait a few minutes and try again.
> 	```bash
> 	dotnet run
> 	```

Reviewer: ```suggestion
1. It's now time to try the console application. Type in the following to run the app:

	```bash
	dotnet run
	```

    If you get an error message the Azure OpenAI resources may not have finished deploying. Wait a couple of minutes and try again.
```

## How it works

> 

Reviewer: ```suggestion
The entire application is contained within the **Program.cs** file. The first several lines of code loads up secrets and configuration values that were set in the `dotnet user-secrets` for you during the application provisioning.

    ```csharp
    // == Retrieve the local secrets saved during the Azure deployment ==========
    var config = new ConfigurationBuilder().AddUserSecrets<Program>().Build();
    string openAIEndpoint = config["AZURE_OPENAI_ENDPOINT"];
    string openAIDeploymentName = config["AZURE_OPENAI_GPT_NAME"];
    string openAiKey = config["AZURE_OPENAI_KEY"];
    
    // == Creating the AIClient ==========
    var endpoint = new Uri(openAIEndpoint);
    var credentials = new AzureKeyCredential(openAiKey);
    ```

```

> The client library is available through NuGet, as the `Azure.AI.OpenAI` package. Using the client we can send and receive requests to an Azure OpenAI service deployed in Azure.

Reviewer: ```suggestion
Our application uses the `Azure.AI.OpenAI` client SDK, which is available on [NuGet](https://www.nuget.org/packages/Azure.AI.OpenAI), to send and receive requests to an Azure OpenAI service deployed in Azure.
```

Once the `OpenAIClient` client is created, we provide more context to the model by adding a system prompt. This instructs the model how you'd like it to act during the conversation.

```csharp
var systemPrompt = 
"""
You are a hiking enthusiast who helps people discover fun hikes in their area. You are upbeat and friendly. 
You introduce yourself when first saying hello. When helping people out, you always ask them 
for this information to inform the hiking recommendation you provide:

1. Where they are located
2. What hiking intensity they are looking for

You will then provide three suggestions for nearby hikes that vary in length after you get that information. 
You will also share an interesting fact about the local nature on the hikes when making a recommendation.
""";

completionOptions.Messages.Add(new ChatRequestSystemMessage(systemPrompt));
```

> 

Reviewer: ```suggestion

Notice that you use the `ChatRequestSystemMessage` class for the system message.

```

> Then we can add a user message to the model, and get the response from the model.

Reviewer: ```suggestion
Then we can add a user message to the model by using the `ChatRequestUserMessage` class. 

To have the model generate a response based off the system prompt and the user request, use the `GetChatCompletionsAsync` function.
```

```csharp
string userGreeting = """
Hi! 
Apparently you can help me find a hike that I will like?
""";

completionOptions.Messages.Add(new ChatRequestUserMessage(userGreeting));
Console.WriteLine($"\n\nUser >>> {userGreeting}");

ChatCompletions response = await openAIClient.GetChatCompletionsAsync(completionOptions);
ChatResponseMessage assistantResponse = response.Choices[0].Message;
Console.WriteLine($"\n\nAI >>> {assistantResponse.Content}");

> completionOptions.Messages.Add(new ChatRequestSystemMessage(assistantResponse.Content)); 

Reviewer: ```suggestion
completionOptions.Messages.Add(new ChatRequestAssisstantMessage(assistantResponse.Content)); 
```

```

> Try it out customize the prompt and message and see how the AI model can help you find a hike that you will like.

Reviewer: ```suggestion
To maintain the chat history or context, make sure you add the response from the model as a `ChatRequestAssistantMessage`.

Customize the system prompt and user message to see how the AI model can help you find a hike that you will like.
```

## Clean up resources

> When you no longer need the sample application or resources, remove the corresponding deployment and all resources.
> ```bash
> azd down
> ```

Reviewer: ```suggestion
When you no longer need the sample application or resources, remove the corresponding deployment and all resources.

```bash
azd down
```
```

## Troubleshooting

> On Windows, you may get an error message: "*postprovision.ps1 is not digitally signed. The script will not execute on the system*" after the deployment. This is cause by the script "postprovision" being executed locally after the deployment to create .NET secret that will be used in the application. To avoid this error, execute the command `Set-ExecutionPolicy -Scope Process -ExecutionPolicy Bypass`. And re-run the `azd up` command.

Reviewer: ```suggestion
On Windows, after running `azd up` you may get the following error message:
> *postprovision.ps1 is not digitally signed. The script will not execute on the system* 

The script **postprovision.ps1** is executed to set the .NET user secrets used in the application. To avoid this error, run the following Powershell command:

    ```powershell
    Set-ExecutionPolicy -Scope Process -ExecutionPolicy Bypass
    ```

Then re-run the `azd up` command.
```

> ## Next steps
> 
> ### More sample code
> 
> - more example coming up soon...
> 
> 
> ### Learn Module
> 
> - [Generate text and conversations with .NET and Azure OpenAI Completions](https://learn.microsoft.com/training/modules/open-ai-dotnet-text-completions/)

Reviewer: ```suggestion
## Next steps

- [Generate text and conversations with .NET and Azure OpenAI Completions](/training/modules/open-ai-dotnet-text-completions/)

```
